hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Balancer.java

```diff
@@ -695,7 +695,7 @@ private long getBlockList() throws IOException {
             // update locations
             for (String datanodeUuid : blk.getDatanodeUuids()) {
               final BalancerDatanode d = datanodeMap.get(datanodeUuid);
-              if (datanode != null) { // not an unknown datanode
+              if (d != null && datanode != null) { // not an unknown datanode
```

Comment:
Allen and I talked about this and although it's not completely clear what's going on when datanode is null, you're change is very likely the right thing to do.

So +1 from me and Allen